cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
+        return std::get<T*>(data_);
+    }
+
+    const T* get_data() const {
+        if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val;
+        } else {
+            return std::get<const T*>(data_);
+        }
+    }
 
-    T* get_pointer() const noexcept {
-        return data_.get();
+    bool has_mutable_data() const {
+        return std::holds_alternative<T*>(data_) && (get_mutable_data() != nullptr);
     }
 
-    std::int64_t get_size() const noexcept {
+    void unique() {
+        if (is_data_owner() || size_ == 0) {
+            return;
+        } else {
+            auto immutable_data = get_data();
+            auto copy_data = new T[size_];
+
+            for (std::int64_t i = 0; i < size_; i++) {
+                copy_data[i] = immutable_data[i];
+            }
+
+            reset(copy_data, size_, [](auto obj){ delete[] obj; });
+        }
+    }
+
+    std::int64_t get_size() const {
         return size_;
     }
 
+    std::int64_t get_capacity() const {
+        return capacity_;
+    }
+
+    bool is_data_owner() const {
+        if (data_owned_ptr_ == nullptr) {
+            return false;
+        } else if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else if (auto ptr_val = std::get_if<const T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else {
+            return false;
+        }
+    }
+
     void reset() {
-        data_.reset();
+        data_owned_ptr_.reset();
+        data_ = std::variant<T*, const T*>();
         size_ = 0;
+        capacity_ = 0;
+    }
+
+    void reset(std::int64_t size) {
+        data_owned_ptr_.reset(new T[size], [](auto obj){ delete[] obj; });
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
+    }
+
+    template <typename Deleter>
+    void reset(T* data, std::int64_t size, Deleter&& deleter) {
+        // TODO: check input parameters
+        data_owned_ptr_.reset(data, std::forward<Deleter>(deleter));
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
+    }
+
+    template <typename U = T*>
+    void reset_not_owning(U data = nullptr, std::int64_t size = 0) {
+        data_ = data;
+        size_ = size;
     }
 
-    const T& operator [](std::int64_t index) const noexcept {
-        return get_pointer()[index];
+    void resize(std::int64_t size) {
+        if (is_data_owner() == false) {
+            throw std::runtime_error("cannot resize array with non-owning data");
+        } else if (size <= 0) {
+            reset_not_owning();
+        } else if (get_capacity() < size) {
+            T* new_data = new T[size];
+            std::int64_t min_size = std::min(size, get_size());
+
+            for (std::int64_t i = 0; i < min_size; i++) {
+                new_data[i] = (*this)[i];
+            }
+
+            reset(new_data, size, [](T* obj){ delete[] obj; });
```

Comment:
`reset` is not marked as `noexcept` (I guess it could throw exception as the STL's shared pointer does), so this is a place where memory leak is possible. Don't need to rush and fix it right now, just keep in mind to address at the production stage. Possible options are `unique_ptr` or [finaly](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e19-use-a-final_action-object-to-express-cleanup-if-no-suitable-resource-handle-is-available).